### PR TITLE
Auto scope closing and openning when a sniper with scope attachment is fired

### PR DESCRIPTION
**Pull request by victorbetoni**
When a sniper with a low fire rate (AWP and M107) is fired, the game automatically closes the scope and opens after N milliseconds (which is determined by the gun fire rate in milliseconds). The game will open the scope again only if the gun is still in player's hand after the N milliseconds. It's more realistic and cooler than just keeping the scope oppened.

### CODE REVIEW

File: crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java [hunk 1]
```diff
@@ -672,6 +681,9 @@ public void toggleRightMouseAction(ILiving<?, ?> living, boolean sendUpdate) {
     if (!this.performingRightMouseAction && living.getEntity().isSprinting()) {
       return;
     }
+    if(this instanceof AimableGun && !((AimableGun) this).canAim()){
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java]: No need for this instance check. Override `toggleRightMouseAction` in `AimableGun` and call the super method if `canAim` is true.

File: crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java [hunk 2]
```diff
@@ -377,6 +370,22 @@ private void tryShoot(ILiving<?, ?> living, ItemStack itemStack) {
         shootSound = this.gunItem.getSilencedShootSound().orElse(shootSound);
       }
 
+      if(this instanceof AimableGun && ((AimableGun) this).isAiming(entity, new ItemStack(gunItem))
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java]: Try and do this without the instance check, perhaps add a method to `DefaultGun` that gets called by `tryShoot` to process the shot and then override that method from `AimableGun`.

**Review by josephtarbit — CHANGES_REQUESTED** [review]
Looks good, however a few code improvements could be made.

**Comment by victorbetoni** [comment]
my IDE added some unnecessary spaces automatically, I'll fix it.

File: crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/AimableGun.java [hunk 3]
```diff
@@ -53,6 +70,40 @@ public float getZoomMultiplier(Entity entity, ItemStack itemStack) {
...
+  protected void processShoot(ILiving<?, ?> living, ItemStack itemStack) {
+    Entity entity = living.getEntity();
+
+    GunItem gunItem = super.getGunItem();
+
+    super.processShoot(living, itemStack);
+
+    if(this.isAiming(entity, itemStack)
+            && gunItem.hasBoltAction()
+            && super.getAttachments().stream().anyMatch(AttachmentItem::isScope)) {
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/AimableGun.java]: Don't need to check for scope attachments.

File: crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/AimableGun.java [hunk 4]
```diff
@@ -53,6 +70,40 @@ public float getZoomMultiplier(Entity entity, ItemStack itemStack) {
...
+
+      toggleRightMouseAction(living, false);
+      canAim = false;
+
+      Timer timer = new Timer();
+      timer.schedule(new TimerTask() {
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/AimableGun.java]: Do this without a timer, I suggest using the `lastShotMs` variable and overriding the tick method to check the time passed.

File: crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java [hunk 5]
```diff
@@ -17,14 +17,13 @@
  */
 package com.craftingdead.core.capability.gun;
 
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Optional;
-import java.util.Random;
-import java.util.Set;
+import java.util.*;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java]: Avoid wildcard imports.

File: crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java [hunk 6]
```diff
@@ -17,14 +17,13 @@
...
-import java.util.Optional;
-import java.util.Random;
-import java.util.Set;
+import java.util.*;
+import java.util.Timer;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
+
+import com.craftingdead.core.item.ModItems;
+import net.minecraft.util.*;
```
Comment (josephtarbit) [inline crafting-dead-core/src/main/java/com/craftingdead/core/capability/gun/DefaultGun.java]: Avoid wildcard imports.

**Review by josephtarbit — CHANGES_REQUESTED** [review]
A few more tweaks

**Pull request merged**